Modifying XATTRs doesnt change the ctime

gaurkuma opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
Changing any metadata, should modify the ctime. 
<!---
Documentation on ZFS Buildbot options can be found at
https://github.com/zfsonlinux/zfs/wiki/Buildbot-Options
-->

### Description
<!--- Describe your changes in detail -->

Default Behaviour
root# stat test.log 
  File: `test.log'
  Size: 32        	Blocks: 9          IO Block: 512    regular file
Device: 18h/24d	Inode: 128         Links: 1
Access: (0600/-rw-------)  Uid: (    0/    root)   Gid: (    0/    root)
Access: 2017-08-29 23:28:04.926223299 -0700
Modify: 2017-08-29 21:51:24.495046598 -0700
Change: 2017-08-29 21:51:24.495046598 -0700

root# setfattr -n user.name1 -v "val1" test.log
root# stat test.log
  File: `test.log'
  Size: 32        	Blocks: 9          IO Block: 512    regular file
Device: 18h/24d	Inode: 128         Links: 1
Access: (0600/-rw-------)  Uid: (    0/    root)   Gid: (    0/    root)
Access: 2017-08-29 23:28:04.926223299 -0700
Modify: 2017-08-29 21:51:24.495046598 -0700
Change: 2017-08-29 21:51:24.495046598 -0700

This patch updates the ctime when xattr is set. Currently the patch only takes care of the case where xattr=sa is set as the property. 

Result:

root# stat test.log 
  File: `test.log'
  Size: 64        	Blocks: 8          IO Block: 4096   regular file
Device: 801h/2049d	Inode: 425987      Links: 1
Access: (0600/-rw-------)  Uid: (    0/    root)   Gid: (    0/    root)
Access: 2017-08-30 16:52:05.519465323 -0700
Modify: 2017-08-30 16:52:05.519465323 -0700
Change: 2017-08-30 16:52:05.519465323 -0700

root# setfattr -n user.name1 -v "val1" test.log 
root# stat test.log 
  File: `test.log'
  Size: 64        	Blocks: 8          IO Block: 4096   regular file
Device: 801h/2049d	Inode: 425987      Links: 1
Access: (0600/-rw-------)  Uid: (    0/    root)   Gid: (    0/    root)
Access: 2017-08-30 16:52:05.519465323 -0700
Modify: 2017-08-30 16:52:05.519465323 -0700
Change: 2017-08-30 16:52:33.220226152 -0700

### Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

### How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
<!--- If your change is a performance enhancement, please provide benchmarks here. -->

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ x] My code follows the ZFS on Linux code style requirements.
- [ ] I have updated the documentation accordingly.
- [ x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [ ] All commit messages are properly formatted and contain `Signed-off-by`.
- [ ] Change has been approved by a ZFS on Linux member.

behlendorf commented on module/zfs/zfs_sa.c:
```diff
@@ -248,9 +248,21 @@ zfs_sa_set_xattr(znode_t *zp)
+		SA_ADD_BULK_ATTR(bulk, count, SA_ZPL_DXATTR(zfsvfs),
+		    NULL, obj, size);
+		/* Any xattr change is metadata change hence update the ctime */
+		zfs_tstamp_update_setup(zp, STATE_CHANGED, NULL, ctime);
```
style: it might be nice to move the `zfs_tstamp_update_setup()` call easier so all the `SA_ADD_BULK_ATTR()` calls are together.

gaurkuma replied:
done

behlendorf commented on module/zfs/zfs_sa.c:
```diff
@@ -248,9 +248,21 @@ zfs_sa_set_xattr(znode_t *zp)
+
 		dmu_tx_commit(tx);
+		/* Update Inode ctime */
+		zfs_inode_update(zp);
```
There's no need to call `zfs_inode_update()`.  The call to `zfs_tstamp_update_setup()` will update the ctime in the inode.  We're working to retire the need for `zfs_inode_update()` so the fewer callers the better.

gaurkuma replied:
done

gaurkuma commented on module/zfs/zfs_sa.c:
```diff
@@ -248,9 +248,21 @@ zfs_sa_set_xattr(znode_t *zp)
-		VERIFY0(sa_update(zp->z_sa_hdl, SA_ZPL_DXATTR(zfsvfs),
-		    obj, size, tx));
+		int count = 0;
+		sa_bulk_attr_t  bulk[2];
```
done

gaurkuma commented on module/zfs/zfs_sa.c:
```diff
@@ -248,9 +248,21 @@ zfs_sa_set_xattr(znode_t *zp)
+
+		SA_ADD_BULK_ATTR(bulk, count, SA_ZPL_DXATTR(zfsvfs),
+		    NULL, obj, size);
+		/* Any xattr change is metadata change hence update the ctime */
```
done

gaurkuma commented:
@behlendorf I have updated the patch. I need some clarity regarding the test coverage. I have added setting xattr and checking if ctime changes in the ctime_001_pos.c file. But as you mentioned we need to test for both xattr=sa and dir. Is this what seems fair to do:

rename ctime_001_pos.c to ctime.c
Add a script ctime_001_pos.ksh
In the script set xattr to dir
Call ctime 
Modify xattr to sa
call ctime

behlendorf commented:
@gaurkuma yes, that sounds entirely reasonable.  Thanks.

gaurkuma commented:
@behlendorf Final Patch Updated.

tuxoko commented on module/zfs/zfs_sa.c:
```diff
@@ -248,8 +248,17 @@ zfs_sa_set_xattr(znode_t *zp)
+		SA_ADD_BULK_ATTR(bulk, count, SA_ZPL_CTIME(zfsvfs),
+		    NULL, &ctime, 16);
+		VERIFY0(sa_bulk_update(zp->z_sa_hdl, bulk, count, tx));
+		zfs_tstamp_update_setup(zp, STATE_CHANGED, NULL, ctime);
```
@gaurkuma you need to move this line before `sa_bulk_update`. Otherwise, you are using uninitialized ctime. You won't see that with stat(2), but you should be able see that with zdb.

gaurkuma replied:
@tuxoko Good Catch..Fixed it.

tuxoko commented on tests/runfiles/linux.run:
```diff
@@ -355,7 +355,7 @@ tests = ['compress_001_pos', 'compress_002_pos', 'compress_003_pos',
 
 [tests/functional/ctime]
-tests = ['ctime_001_pos' ]
+tests = ['ctime' ]
```
I think it would be better to switch the two file name, so we get more consistent with others.

gaurkuma replied:
@tuxoko The reason I didnt do this way was because ctime_001_pos is a .c file and renaming it will mean changing multiple places in the Makefile. If that's ok, I will go ahead and make the change.

behlendorf replied:
That's fine, please go ahead and rename it.

gaurkuma commented:
@tuxoko Updated the patch..

behlendorf commented on tests/zfs-tests/tests/functional/ctime/Makefile.am:
```diff
@@ -3,10 +3,11 @@ include $(top_srcdir)/config/Rules.am
-pkgexec_PROGRAMS = ctime_001_pos
-ctime_001_pos_SOURCES = ctime_001_pos.c
+pkgexec_PROGRAMS = ctime
+ctime_001_pos_SOURCES = ctime.c
```
Shouldn't this now be `ctime_SOURCES = ctime.c`?

gaurkuma replied:
@behlendorf ..Done

behlendorf commented:
@tonyhutter this is a good candidate for 0.7, can you open a PR to port it.  I expect it'll apply cleanly.
